packages/block-editor/src/components/inserter/block-list.js

```diff
@@ -0,0 +1,462 @@
+				}
+			}
+			const destinationRootBlockName = getBlockName(
+				destinationRootClientId
+			);
+
+			const { __experimentalFetchReusableBlocks } = getSettings();
+
+			return {
+				categories: getCategories(),
+				collections: getCollections(),
+				rootChildBlocks: getChildBlockNames( destinationRootBlockName ),
+				items: getInserterItems( destinationRootClientId ),
+				destinationRootClientId: destRootClientId,
+				fetchReusableBlocks: __experimentalFetchReusableBlocks,
+				...pick( select( 'core/block-editor' ), [
+					'getSelectedBlock',
+					'getBlockIndex',
+					'getBlockSelectionEnd',
+					'getBlockOrder',
+				] ),
+			};
+		},
+		[ clientId, isAppender, rootClientId ]
+	);
+	const {
+		replaceBlocks,
+		insertBlock,
+		showInsertionPoint,
+		hideInsertionPoint,
+	} = useDispatch( 'core/block-editor' );
+
+	// Fetch resuable blocks on mount
+	useEffect( () => {
+		fetchReusableBlocks();
+	}, [] );
+
+	// To avoid duplication, getInsertionIndex is extracted and used in two event handlers
+	// This breaks the withDispatch not containing any logic rule.
+	// Since it's a function only called when the event handlers are called,
+	// it's fine to extract it.
+	// eslint-disable-next-line no-restricted-syntax
+	function getInsertionIndex() {
+		// If the clientId is defined, we insert at the position of the block.
+		if ( clientId ) {
+			return getBlockIndex( clientId, destinationRootClientId );
+		}
+
+		// If there a selected block, we insert after the selected block.
+		const end = getBlockSelectionEnd();
+		if ( ! isAppender && end ) {
+			return getBlockIndex( end, destinationRootClientId ) + 1;
+		}
+
+		// Otherwise, we insert at the end of the current rootClientId
+		return getBlockOrder( destinationRootClientId ).length;
+	}
+
+	const onHoverItem = ( item ) => {
+		onHover();
+		if ( item ) {
+			const index = getInsertionIndex();
+			showInsertionPoint( destinationRootClientId, index );
+		} else {
+			hideInsertionPoint();
+		}
+	};
+
+	const onSelectItem = ( item ) => {
+		const { name, title, initialAttributes, innerBlocks } = item;
+		const selectedBlock = getSelectedBlock();
+		const insertedBlock = createBlock(
+			name,
+			initialAttributes,
+			createBlocksFromInnerBlocksTemplate( innerBlocks )
+		);
+		if (
+			! isAppender &&
+			selectedBlock &&
+			isUnmodifiedDefaultBlock( selectedBlock )
+		) {
+			replaceBlocks( selectedBlock.clientId, insertedBlock );
+		} else {
+			insertBlock(
+				insertedBlock,
+				getInsertionIndex(),
+				destinationRootClientId,
+				selectBlockOnInsert
+			);
+
+			if ( ! selectBlockOnInsert ) {
+				// translators: %s: the name of the block that has been added
+				const message = sprintf( __( '%s block added' ), title );
+				speak( message );
+			}
+		}
+
+		onSelect();
+		return insertedBlock;
+	};
+
+	const filteredItems = useMemo( () => {
+		return searchItems( items, categories, collections, filterValue );
```

Comment:
The chances of reuse the return value is very low because of the dynamic nature of the `filterValue` param. Are you sure the use of memoization is still beneficial?
I see other memoized computations based on the `filteredItems` so maybe it all makes sense after all. I guess `useMemo` stores only 1 item so it could bring some benefits in cases when other props change.

Reply:
Yeah, it seems like this component can rerender for several reasons (store, props) so it does have some value. The important part here is that `filteredItems` is used by other hooks to compute other variables and all these filtering/grouping can be costly.

Reply:
Yes, I figured it out myself later when scanning the rest of the component 👍